### ISSUE TITLE
invalid memory address or nil pointer dereference azure.mapFromVMScaleSetVM

### ISSUE THREAD

**Issue by McHammar** [prometheus/prometheus#13574]
### What did you do?

The prometheus pod can't start anymore and gets this error message before dying

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x1211d1b]
goroutine 762 [running]:
github.com/prometheus/prometheus/discovery/azure.mapFromVMScaleSetVM({0xc0447b9710, 0x0, 0x0, 0x0, 0x0, 0xc0447b9740, 0xc0447b9720, 0xc0447b9730, {0x0, 0x0, ...}, ...}, ...)
	/app/discovery/azure/azure.go:609 +0x5b
github.com/prometheus/prometheus/discovery/azure.(*azureClient).getScaleSetVMs(0xc04b804540, {0x3f6e318, 0xc04abdf7c0}, {0xc07b482290, 0x0, 0x0, 0x0, 0xc0796e28c0, 0xc07b1e5410, 0x0, ...})
	/app/discovery/azure/azure.go:562 +0x49b
github.com/prometheus/prometheus/discovery/azure.(*Discovery).refresh(0xc04ab9f890, {0x3f6e318?, 0xc04abdf7c0})
	/app/discovery/azure/azure.go:354 +0x8b0
github.com/prometheus/prometheus/discovery/refresh.(*Discovery).refresh(0xc04ab93a40, {0x3f6e318, 0xc04abdf7c0})
	/app/discovery/refresh/refresh.go:121 +0xcf
github.com/prometheus/prometheus/discovery/refresh.(*Discovery).Run(0xc04ab93a40, {0x3f6e318, 0xc04abdf7c0}, 0xc04a2bf140)
	/app/discovery/refresh/refresh.go:77 +0x66
created by github.com/prometheus/prometheus/discovery/legacymanager.(*Manager).startProvider in goroutine 107
	/app/discovery/legacymanager/manager.go:200 +0x3a5
```

If I remove all Azure sd configs the pod is able to start again

### What did you expect to see?

Pod starting

### What did you see instead? Under which circumstances?

Crashloop

### System information

AKS 1.27.3

### Prometheus version

```text
v2.49.1
```

**Comment by darshanime** [comment]
from the stacktrace, the issue seems to be [here](https://github.com/prometheus/prometheus/blob/v2.49.1/discovery/azure/azure.go#L609). let me raise a PR

```diff
 func mapFromVMScaleSetVM(vm armcompute.VirtualMachineScaleSetVM, scaleSetName string) virtualMachine {
-	osType := string(*vm.Properties.StorageProfile.OSDisk.OSType)
+	osType := getOSTypeFromVirtualMachineScaleSetVM(vm)
 	tags := map[string]*string{}
 	networkInterfaces := []string{}
 	var computerName string
@@ -702,3 +702,23 @@ func (d *Discovery) getFromCache(nicID string) (*armnetwork.Interface, bool) {
 	net, found := d.cache.Get(nicID)
 	return net, found
 }
+
+func getOSTypeFromVirtualMachineScaleSetVM(vm armcompute.VirtualMachineScaleSetVM) string {
+	if vm.Properties == nil {
+		return ""
+	}
+
+	if vm.Properties.StorageProfile == nil {
+		return ""
+	}
+
+	if vm.Properties.StorageProfile.OSDisk == nil {
+		return ""
+	}
+
+	if vm.Properties.StorageProfile.OSDisk.OSType == nil {
+		return ""
+	}
+
+	return string(*vm.Properties.StorageProfile.OSDisk.OSType)
+}

```

**Comment by bboreham** [comment]
Closed by #13578

**Comment by bboreham** [comment]
I think it's more like #13643, which was a bad merge, fixed by #13702.

There is a release candidate for 2.51.0 out now which has this fix.
